### PR TITLE
Annotate CommandLineParser/Arguments and File/PathUtilities

### PR DESCRIPTION
**Pull request by jcouv**
This found a bug (added tests and a simple fix in `CommandLineParser.EnumerateFiles`).
I also added a null check in `CommandLineParser` (line 428) which I could not protect or rationalize otherwise.

### CODE REVIEW

File: src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs [hunk 1]
```diff
@@ -73,72 +75,70 @@ public abstract class CommandLineArguments
...
         /// Compilation name or null if not specified.
         /// </summary>
-        public string CompilationName { get; internal set; }
+        public string? CompilationName { get; internal set; }
 
         /// <summary>
         /// Gets the emit options.
         /// </summary>
-        public EmitOptions EmitOptions { get; internal set; }
+        public EmitOptions EmitOptions { get; internal set; } = null!; // initialized by Parse
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs]: ❓ Why not make this a constructor parameter (required initialization)?

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 2]
```diff
@@ -147,7 +149,7 @@ private static string GetDirectoryName(string path, bool isUnixLike)
             return null;
         }
 
-        internal static bool IsSameDirectoryOrChildOf(string child, string parent)
+        internal static bool IsSameDirectoryOrChildOf([DisallowNull]string? child, string parent)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]:
```suggestion
        internal static bool IsSameDirectoryOrChildOf([NotNullWhen(true)] string? child, string parent)
```

File: src/Compilers/Core/Portable/RuleSet/RuleSet.cs [hunk 3]
```diff
@@ -287,7 +287,8 @@ public static ReportDiagnostic GetDiagnosticOptionsFromRulesetFile(string rulese
             return GetDiagnosticOptionsFromRulesetFile(rulesetFileFullPath, out specificDiagnosticOptions, null, null);
         }
 
-        internal static ReportDiagnostic GetDiagnosticOptionsFromRulesetFile(string rulesetFileFullPath, out Dictionary<string, ReportDiagnostic> diagnosticOptions, IList<Diagnostic> diagnosticsOpt, CommonMessageProvider messageProviderOpt)
+#nullable enable
+        internal static ReportDiagnostic GetDiagnosticOptionsFromRulesetFile(string? rulesetFileFullPath, out Dictionary<string, ReportDiagnostic> diagnosticOptions, IList<Diagnostic> diagnosticsOpt, CommonMessageProvider messageProviderOpt)
```
Comment (sharwell) [inline src/Compilers/Core/Portable/RuleSet/RuleSet.cs]:
```suggestion
        internal static ReportDiagnostic GetDiagnosticOptionsFromRulesetFile(string? rulesetFileFullPath, out Dictionary<string, ReportDiagnostic> diagnosticOptions, IList<Diagnostic>? diagnosticsOpt, CommonMessageProvider? messageProviderOpt)
```

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: I don't want callers to pass `null`. 
`[NotNullWhen(true)]` is a post-condition only.

---
In reply to: [368191535](https://github.com/dotnet/roslyn/pull/41063#discussion_r368191535) [](ancestors = 368191535)

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 4]
```diff
@@ -1894,11 +1900,12 @@ private static IEnumerable<InstrumentationKind> ParseInstrumentationKinds(string
                 return null;
             }
 
-            if (string.IsNullOrEmpty(filePath))
+            if (RoslynString.IsNullOrWhiteSpace(filePath))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: :memo: this change aligns with logic in `ParseResourceDescription` (called a few lines above). The effect is to change which error is reported (see `ParseResources` test, which was updated).

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 5]
```diff
@@ -1316,18 +1318,22 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
...
-            if (!string.IsNullOrEmpty(baseDirectory))
+            if (!RoslynString.IsNullOrEmpty(baseDirectory))
             {
                 keyFileSearchPaths.Add(baseDirectory);
             }
 
-            if (!string.IsNullOrEmpty(outputDirectory) && baseDirectory != outputDirectory)
+            if (RoslynString.IsNullOrEmpty(outputDirectory))
+            {
+                AddDiagnostic(diagnostics, ErrorCode.ERR_NoOutputDirectory);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: :memo: this new error guards against `null` output directory, allowing us to declare `OutputDirectory` as non-nullable.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler @sharwell for review. This annotates more of the public APIs. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler @sharwell for review. This annotates more of the public APIs. Thanks

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: 📝 This new error means loc change, and the cutoff for 16.5 already passed so this needs to retarget 16.6 Preview 1

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 6]
```diff
@@ -1708,9 +1714,9 @@ private static OutputKind ParseTarget(string value, IList<Diagnostic> diagnostic
             }
         }
 
-        private static IEnumerable<string> ParseUsings(string arg, string value, IList<Diagnostic> diagnostics)
+        private static IEnumerable<string> ParseUsings(string arg, string? value, IList<Diagnostic> diagnostics)
         {
-            if (value.Length == 0)
+            if (value is null || value.Length == 0)
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]:
```suggestion
            if (RoslynString.IsNullOrEmpty(value))
```

File: src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs]: Tried it after your suggestion, but decided against it as it breaks the regularity of the code in parser and also implies a change in VB code. Neither of these seem worth it.
The topic of field initializers will be re-discussed in LDM and at the very least we'll provide some guidance to the recommended pattern for dealing with this situation.

---
In reply to: [368190989](https://github.com/dotnet/roslyn/pull/41063#discussion_r368190989) [](ancestors = 368190989)

**Comment by jcouv** [comment]
@sharwell I addressed your feedback. I'll retarget to 16.6p1 after the review is complete. Thanks

**Comment by jcouv** [comment]
@333fred @agocke @jaredpar for review (nullable annotations). Thanks

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 7]
```diff
@@ -115,12 +118,12 @@ public static string GetFileName(string path, bool includeExtension = true)
...
         /// </remarks>
         /// <returns>Prefix of path that represents a directory</returns>
-        public static string GetDirectoryName(string path)
+        public static string? GetDirectoryName(string? path)
         {
             return GetDirectoryName(path, IsUnixLikePlatform);
         }
 
-        private static string GetDirectoryName(string path, bool isUnixLike)
+        private static string? GetDirectoryName(string? path, bool isUnixLike)
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: Should be `NotNullIfNotNull("path")`, and above.

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 8]
```diff
@@ -149,7 +152,7 @@ private static string GetDirectoryName(string path, bool isUnixLike)
             return null;
         }
 
-        internal static bool IsSameDirectoryOrChildOf(string child, string parent)
+        internal static bool IsSameDirectoryOrChildOf([DisallowNull]string? child, string parent)
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: This would be much clearer if you made a temp `string?` variable. This exposes private implementation details of `IsSameDirectoryOrChildOf` to callers, which seems like bad practice.

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 9]
```diff
@@ -171,12 +174,13 @@ internal static bool IsSameDirectoryOrChildOf(string child, string parent)
         /// <summary>
         /// Gets the root part of the path.
         /// </summary>
-        public static string GetPathRoot(string path)
+        public static string? GetPathRoot(string? path)
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: This should also be `NotNullIfNotNull`

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 10]
```diff
@@ -335,12 +339,10 @@ public static PathKind GetPathKind(string path)
             return PathKind.Relative;
         }
 
-#nullable enable
-
         /// <summary>
         /// True if the path is an absolute path (rooted to drive or network share)
         /// </summary>
-        public static bool IsAbsolute([NotNullWhen(true)]string? path)
+        public static bool IsAbsolute([NotNullWhen(returnValue: true)] string? path)
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: >returnValue: [](start = 51, length = 12)

Why add this? Seems unnecessary

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 11]
```diff
@@ -421,12 +421,12 @@ public static string CombinePossiblyRelativeAndRelativePaths(string rootOpt, str
                     return null;
             }
 
-            return CombinePathsUnchecked(rootOpt, relativePath);
+            return CombinePathsUnchecked(rootOpt, relativePath!);
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: >! [](start = 62, length = 1)

Doesn't seem like a suppression should be needed here.

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 12]
```diff
@@ -393,7 +393,7 @@ public static string CombineAbsoluteAndRelativePaths(string root, string relativ
         {
             Debug.Assert(IsAbsolute(root));
 
-            return CombinePossiblyRelativeAndRelativePaths(root, relativePath);
+            return CombinePossiblyRelativeAndRelativePaths(root, relativePath)!;
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: I don't see how this is safe?

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 13]
```diff
@@ -703,11 +703,11 @@ public static string NormalizePathPrefix(string filePath, ImmutableArray<KeyValu
         /// 
         /// The more accurate way is to let the framework parse the path and throw on any errors.
         /// </summary>
-        public static bool IsValidFilePath(string fullPath)
+        public static bool IsValidFilePath([NotNullWhen(returnValue: true)] string? fullPath)
```
Comment (333fred) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: >returnValue: [](start = 56, length = 12)

Same comment.

File: src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs [hunk 14]
```diff
@@ -464,7 +465,8 @@ public IEnumerable<AnalyzerReference> ResolveAnalyzerReferences(IAnalyzerAssembl
             EventHandler<AnalyzerLoadFailureEventArgs> errorHandler = (o, e) =>
             {
                 var analyzerReference = o as AnalyzerFileReference;
-                DiagnosticInfo diagnostic;
+                RoslynDebug.Assert(analyzerReference is object);
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs]: Why not just make the `as` cast above a real cast?

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs [hunk 15]
```diff
@@ -53,13 +56,18 @@ internal virtual TextReader CreateTextFileReader(string fullPath)
         /// <param name="fileNamePattern">File name pattern. May contain wildcards '*' (matches zero or more characters) and '?' (matches any character).</param>
         /// <param name="searchOption">Specifies whether to search the specified <paramref name="directory"/> only, or all its subdirectories as well.</param>
         /// <returns>Sequence of file paths.</returns>
-        internal virtual IEnumerable<string> EnumerateFiles(string directory, string fileNamePattern, SearchOption searchOption)
+        internal virtual IEnumerable<string> EnumerateFiles(string? directory, string fileNamePattern, SearchOption searchOption)
         {
+            if (directory is null)
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: What prompted this change?

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs [hunk 16]
```diff
@@ -426,9 +430,10 @@ internal static string RemoveTrailingSpacesAndDots(string path)
                             }
                         }
 
-                        if (responsePaths != null)
+                        if (responsePaths != null &&
+                            PathUtilities.GetDirectoryName(resolvedPath) is string directory)
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: >PathUtilities.GetDirectoryName(resolvedPath) is string directory [](start = 28, length = 64)

Why add this condition and not just assert the result is not null in the body of the `if`? This potentially changes evaluation semantics.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: What was the previous behavior?

---
In reply to: [370462374](https://github.com/dotnet/roslyn/pull/41063#discussion_r370462374) [](ancestors = 370462374)

**Comment by 333fred** [comment]
Done review pass (commit 11)

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: You could get a null OutputDirectory, which we're not prepared to handle in code that reads that property.
I'm not sure if that emitting code (which reads the property) could be reached though.

---
In reply to: [377830798](https://github.com/dotnet/roslyn/pull/41063#discussion_r377830798) [](ancestors = 377830798,370462374)

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: That would make things so much easier, but I don't think that's possible. 
I see a path where `path` is not null, but we still return null from the method

---
In reply to: [377802189](https://github.com/dotnet/roslyn/pull/41063#discussion_r377802189) [](ancestors = 377802189)

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: We use it in most other places. I don't much care though

---
In reply to: [377805204](https://github.com/dotnet/roslyn/pull/41063#discussion_r377805204) [](ancestors = 377805204)

File: src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs (hunk 14 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs]: Didn't want to change code, in case my understanding of the code was wrong (changing to a cast would throw).

---
In reply to: [377819607](https://github.com/dotnet/roslyn/pull/41063#discussion_r377819607) [](ancestors = 377819607)

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs (hunk 15 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: IsAbsolute returns false on `null`, so assertion below fails

---
In reply to: [377820068](https://github.com/dotnet/roslyn/pull/41063#discussion_r377820068) [](ancestors = 377820068)

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs (hunk 16 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: If we did that, I think we'd hit the assertion. I'm not 100% though (didn't manage to hit it in test, but this may be platform-dependent...)

---
In reply to: [377822593](https://github.com/dotnet/roslyn/pull/41063#discussion_r377822593) [](ancestors = 377822593)

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs (hunk 12 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/FileSystem/PathUtilities.cs]: I'm not sure what I was thinking. Will remove. Thanks

---
In reply to: [377806395](https://github.com/dotnet/roslyn/pull/41063#discussion_r377806395) [](ancestors = 377806395)

File: src/Compilers/Core/Portable/AssemblyUtilities.cs [hunk 17]
```diff
@@ -85,8 +82,7 @@ public static ImmutableArray<string> FindAssemblySet(string filePath)
         /// <exception cref="BadImageFormatException">If the file is not an assembly or is somehow corrupted.</exception>
         public static Guid ReadMvid(string filePath)
         {
-            Debug.Assert(filePath != null);
-            Debug.Assert(PathUtilities.IsAbsolute(filePath));
+            RoslynDebug.Assert(PathUtilities.IsAbsolute(filePath));
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/AssemblyUtilities.cs]: Why is this necessary? Now that we're targeting `netcoreapp3.1` there shouldn't be any reason to use `RoslynDebug` over `Debug` that I'm aware of.

File: src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs [hunk 18]
```diff
@@ -75,72 +77,70 @@ public abstract class CommandLineArguments
         /// <summary>
         /// Compilation name or null if not specified.
         /// </summary>
-        public string CompilationName { get; internal set; }
+        public string? CompilationName { get; internal set; }
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommandLineArguments.cs]: I distinctly remember fixing a bug that came about specifically because the code didn't realize `CompilationName` could be `null`. Seeing this makes me happy for the time I spent fixing that bug some years ago.

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs [hunk 19]
```diff
@@ -226,7 +230,7 @@ internal static bool TryParseOption(string arg, out string name, out string valu
         /// <summary>
         /// Trims all '.' and whitespace from the end of the path
         /// </summary>
-        internal static string RemoveTrailingSpacesAndDots(string path)
+        internal static string? RemoveTrailingSpacesAndDots(string? path)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: Did you consider using `NotNullWhenNotNull` to mark the return?

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs [hunk 20]
```diff
@@ -426,9 +430,10 @@ internal static string RemoveTrailingSpacesAndDots(string path)
                             }
                         }
 
-                        if (responsePaths != null)
+                        if (responsePaths != null &&
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: Feels like we're hiding a bug here. Agree this change is necessary because the old code path would essentially throw in the case where `GetDirectoryName` returned `null`. At the same time though now this will just silently continue. Should we issue the diagnostic `FTL_InvalidInputFileName` when we can't get a directory name here?

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs [hunk 21]
```diff
@@ -126,7 +130,7 @@ private static string ResolveRelativePath(PathKind kind, string path, string bas
                         return null;
                     }
 
-                    if (path.Length == 1)
+                    if (path!.Length == 1)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Seems like `path` could be `null` in the case here where `kind == PathKind.Empty`

**Comment by jaredpar** [comment]
Done with review pass (iteration 12)

File: src/Compilers/Core/Portable/AssemblyUtilities.cs (hunk 17 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/AssemblyUtilities.cs]: This PR was started before you did the multi-targeting ;-)
Do you want me to remove all uses of RoslynDebug, or should we do that separately?

---
In reply to: [378074644](https://github.com/dotnet/roslyn/pull/41063#discussion_r378074644) [](ancestors = 378074644)

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 21 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: See assertion I added above. In this branch, we know that `GetPathKind(path) == kind == PathKind.RelativeToCurrentDirectory`

---
In reply to: [378078393](https://github.com/dotnet/roslyn/pull/41063#discussion_r378078393) [](ancestors = 378078393)

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs (hunk 20 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: Makes sense.
I'll add the diagnostic. I've not been able to hit this case yet though, so it will lack coverage...

---
In reply to: [378077429](https://github.com/dotnet/roslyn/pull/41063#discussion_r378077429) [](ancestors = 378077429)

File: src/Compilers/Core/Portable/AssemblyUtilities.cs (hunk 17 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/AssemblyUtilities.cs]: Separately. 

I was more trying to work out how this could impact our flow analysis. Eventually I realized it was because it's `PathUtilities` (a type we define) not `Path` a type defined in SDK. Hence there was actual nullable information involved in this change.

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs [hunk 22]
```diff
@@ -566,6 +566,26 @@ public void RelativePath()
             }
         }
 
+        [Theory]
+        [InlineData(null, null)]
+        [InlineData("c:", null)]
+        [InlineData("c:\\", null)]
```
Comment (cston) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: >"c:\\" [](start = 20, length = 6)

Are these tests only run on Windows?

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs [hunk 23]
```diff
@@ -566,6 +566,26 @@ public void RelativePath()
...
+        [InlineData("c:\\first\\second", "c:\\first")]
+        [InlineData("c:\\first\\second\\", "c:\\first\\second")]
+        [InlineData("c:\\first\\second\\third", "c:\\first\\second")]
+        [InlineData("\\", null)]
+        [InlineData("\\first", "\\")]
+        [InlineData("\\first\\", "\\first")]
+        [InlineData("\\first\\second", "\\first")]
+        [InlineData("\\first\\second\\", "\\first\\second")]
+        [InlineData("\\first\\second\\third", "\\first\\second")]
+        public void TestGetDirectoryName_Windows(string path, string expectedOutput)
```
Comment (cston) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: >public [](start = 8, length = 6)

Do we need to test relative paths as well?

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs (hunk 22 above)
Comment (jcouv) [reply src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: I thought the logic in the method didn't depend on the current platform, but it looks like it does (directory separator is `\` or `/`). I'll make the test conditional.

---
In reply to: [378585981](https://github.com/dotnet/roslyn/pull/41063#discussion_r378585981) [](ancestors = 378585981)

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs (hunk 23 above)
Comment (jcouv) [reply src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: Doesn't hurt. I'll add.
The main point of this test is to show when we're returning null

---
In reply to: [378586494](https://github.com/dotnet/roslyn/pull/41063#discussion_r378586494) [](ancestors = 378586494)

**Comment by 333fred** [comment]
>                         string referencePath = Path.Combine(directory!, referenceName + ".dll");

Consider using an assert instead of a suppression.

---
Refers to: src/Compilers/Core/Portable/AssemblyUtilities.cs:63 in 886d453. [](commit_id = 886d45369a0b1b034e658c87881b32ffa6dc79db, deletion_comment = False)

File: src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs [hunk 24]
```diff
@@ -329,22 +334,23 @@ internal static string RemoveTrailingSpacesAndDots(string path)
             }
             else
             {
-                pdbPath = Path.ChangeExtension(Path.Combine(outputDirectory, outputFileName), ".pdb");
+                // If outputDirectory were null, then outputFileName would be null (see ParseAndNormalizeFile)
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/CommandLineParser.cs]: These comments are better served with an assert for verification.

**Comment by jcouv** [comment]
>                         string referencePath = Path.Combine(directory!, referenceName + ".dll");

Hum, this is an existing suppression, and I'm not sure it's justified... I'll add to tracking issue for null directories

---
In reply to: [586411159](https://github.com/dotnet/roslyn/pull/41063#issuecomment-586411159) [](ancestors = 586411159)

---
Refers to: src/Compilers/Core/Portable/AssemblyUtilities.cs:63 in 886d453. [](commit_id = 886d45369a0b1b034e658c87881b32ffa6dc79db, deletion_comment = False)

**Comment by jaredpar** [comment]
@jcouv @333fred 

> Hum, this is an existing suppression, and I'm not sure it's justified...

It's justified. The rational is captured here  https://github.com/dotnet/roslyn/pull/41485#discussion_r376601828

**Comment by jaredpar** [comment]
@jcouv @333fred 

At the same time though several people have asked about why this suppression exists. Clearly the context is unclear. At the least I should add a comment here to justify why this is done this way.

**Comment by jcouv** [comment]
@jaredpar I’ll update the comment

**Comment by jcouv** [comment]
@jaredpar I'm going to leave the comment as-is for follow-up. Reading the rationale again, I suspect that it is incorrect. If you open a file in the root directory, the `OpenRead` would succeed, but `GetDirectoryName` would return `null` (that's really annoying behavior in top-level directory in my opinion, but that seems to be the case).

I'll go ahead and merge leaving the follow-up comment.
